include/session/attachments.hpp

```diff
@@ -0,0 +1,208 @@
+///
+/// If a problem with the data is found, the `update()` or `finalize()` call will returns false
+/// indicating that the decryption failed, and any partially decrypted output data provided to the
+/// output callback should be discarded or deleted.  Further calls to `update()` or `finalize()`
+/// after such a failure will simply return false without processing any additional data.
+///
+/// This class is not recommended if the intention is to build an in-memory buffer from existing
+/// in-memory data: `decrypt()` will be more memory efficient in that case.
+class Decryptor {
+    std::function<void(std::span<const std::byte> decrypted)> output;
+    std::vector<std::byte> buf;
+    bool header = false;
+    bool depadded = false;
+    bool failed = false;
+    bool finished = false;
+    bool hit_final = false;
+    cleared_uc32 key;
+    unsigned char st_data[52];  // crypto_secretstream_xchacha20poly1305_state data
+
+    void process_header(std::span<const std::byte, 1 + ENCRYPT_HEADER> chunk);
+    void process_chunk(std::span<const std::byte> chunk, bool is_final = false);
+
+  public:
+    /// Constructs a decryptor.  The given output will be called as soon as enough data has been
+    /// accumulated to validate additional decrypted data.
+    Decryptor(
+            std::span<const std::byte, ENCRYPT_KEY_SIZE> key,
+            std::function<void(std::span<const std::byte> decrypted)> output);
+
+    /// Provides more data to the decryptor.  If the additional data completes an input data chunk
+    /// then output will be called with the partially decrypted data.  Returns true if the data was
+    /// accepted, false if data stream decryption failed (either because of the new data, or some
+    /// previous failure).
+    ///
+    /// Throws std::logic_error if called after a successful finalize().
+    bool update(std::span<const std::byte> enc_data);
+
+    /// Called to signal the end of the encrypted data stream.  If all data was processed
+    /// successfully and the stream ended properly, this returns true; returns false if the stream
+    /// data did not indicate finality (or if a previous update returned failure).
+    ///
+    /// Throws std::logic_error if called after a successful finalize().
+    bool finalize();
+};
+
+/// API: crypto/attachment::decrypt
+///
+/// Decrypts an attachment allegedly produced by attachment::encrypt to an output file.  Overwrites
+/// the file if it already exists.
+///
+/// Inputs:
+/// - `data` -- in-memory buffer of data to decrypt.
+/// - `key` -- the 32-byte decryption key.
+/// - `filename` -- where to write the output file.
+///
+/// Outputs: None.
+///
+/// Throws std::runtime_error if decryption fails or if writing to the file fails.  Upon exception a
+/// partially written file will be deleted.
+void decrypt(
+        std::span<const std::byte> encrypted,
+        std::span<const std::byte, ENCRYPT_KEY_SIZE> key,
+        const std::filesystem::path& filename);
```

Comment:
Would it be beneficial to have a version of this which takes the file path of an encrypted file and an output path so clients don't need to use `Decryptor` directly for a file->file decryption?

Reply:
There is now a file->file version of `decrypt` (along with a bunch of other alternatives).